README.md
# Laravel Scoring

<p align="center">
    <a href="https://github.styleci.io/repos/402087252"><img src="https://github.styleci.io/repos/402087252/shield?style=flat" alt="StyleCI Status"></a>
    <a href="https://packagist.org/packages/arttiger/laravel-scoring"><img src="https://img.shields.io/packagist/dt/arttiger/laravel-scoring?style=flat" alt="Total Downloads"></a>
    <a href="https://packagist.org/packages/arttiger/laravel-scoring"><img src="https://img.shields.io/packagist/v/arttiger/laravel-scoring?style=flat" alt="Latest Stable Version"></a>
    <a href="https://packagist.org/packages/arttiger/laravel-scoring"><img src="https://img.shields.io/packagist/l/arttiger/laravel-scoring?style=flat" alt="License"></a>
</p>

Integration of scoring systems for Laravel. The package provides integration scoring systems:
- RiskTools

## Installation

Install the package via composer:

```
composer require arttiger/laravel-scoring
```

### Configuration

In order to edit the default configuration you may execute:

```
php artisan vendor:publish --provider="Arttiger\Scoring\ScoringServiceProvider" --tag="config"
```

After that, `config/scoring.php` will be created.

### Environment

Set environment variable (`.env`)

```
RISK_TOOLS_URL=
RISK_TOOLS_KEY=
```

## Usage

#### Get scoring data

You can get data from the scoring system as follows:

```
$result = Scoring::scoring('risk_tools', $attributes);
```

- 'risk_tools' - scoring system;
- $attributes - data (array) from the credit model according to the mapping of variables in `config/scoring.php`. This map establishes the correspondence between the attributes of your model and the required query fields in scoring system.

```
'model_data' => [
    // For "scoring" method
    'id'              => 'public_id',     // ID of loan
    'user_id'         => 'client_id',     // ID of client
    'social_number'   => 'inn',           // INN of client
    'phone'           => 'phone',         // Phone of client
    'email'           => 'email',         // Email of client
    'passport_number' => 'passport',      // Passport (example АА123456 or number ID-card)
    'passport_date'   => 'passport_date', // Passport issue date (example "2001-12-01")
    'first_name'      => 'firstName',     // First name
    'last_name'       => 'lastName',      // Last name
    'other_name'      => 'middleName',    // Middle name
    'birth_date'      => 'birth_date',    // Birth date (example "2001-12-01")
    'gender_id'       => 'gender_id',     // Gender (1 - male, 2 - female)
    'loan_amount'     => 'sum',           // Amount of credit
    'loan_days'       => 'period',        // Number days of credit
    'applied_at'      => 'created_at',    // Date and time created loan (example "2018-04-05T19:29:51+03:00")
    'ip'              => 'ip',            // IP of client
    'user_agent'      => 'user_agent',    // User-Agent of browser of client
    'ubki'            => 'ubki',          // XML-data from UBKI

    // For updating of statuses
    'status_id'       => 'status_code',   // Status code of credit (1 - NEW, 2 - REJECT, 3 - APPROVED, 4 - ISSUED, 5 - CLOSED, 6 - OVERDUE)
    'closed_at'       => 'closed_at',     // Date and time close loan (example "2018-04-25T10:05:00+03:00")
    'amount_to_pay'   => 'amount_debt',   // Total debt
    'total_paid'      => 'total_paid',    // The sum of all payments
    'overdue_days'    => 'overdue_days',  // The total number of days of delay of loan
],
```

`$result` - response from scoring system (array).

```
$result = [
    "group" => 4,
    "score" => 0.495,
    "amount_limit_max" => 1200,
    "amount_limit" => 1000.0,
    "filters" => [],
    "score_model" => [
        "version" => "v1_7"
    ]
]
```

#### Send status of credit

Send status of credit to the scoring system as follows:

```
$result = Scoring::status('risk_tools', $attributes);
```

- 'risk_tools' - scoring system;
- $attributes - data (array) from the credit model according to the mapping of variables

#### Get the UBKI reports

Get the UBKI reports from the scoring system as follows:

```
$result = Scoring::ubkiReport('risk_tools', ['public_id' => $public_id]);
```

- $public_id - ID of credit

```
$result = [
    "reports": [
        [
            "report_date" => "2018-12-25",
            "xml" => "<xml>"
        ],
        [
            "report_date" => "2018-12-25",
            "xml" => "<xml>"
        ]
    ]
]
```

#### Get pre-scoring data

You can get data from the pre-scoring system as follows:

```
$result = Scoring::pre_scoring('risk_tools', ['social_number' => $social_number]);
```

- $social_number - Social number (INN) of client

#### Trait

Add `ScoringTrait`-trait to the model with client data:

```
    use Arttiger\Scoring\Traits\ScoringTrait;

    class Loan extends Model
    {
        use ScoringTrait;
        ...
    }
```

Add a new method `scoringAttributes()` to the class to add the necessary attributes and fill them with data:

```
    use Arttiger\Scoring\Traits\ScoringTrait;

    class Loan extends Model
    {
        use ScoringTrait;
        ...
        
        public function scoringAttributes()
        {
            $client_data = json_decode($this->attributes['client_data']);
            $this->attributes['client_id']  = trim($client_data->id); 
            $this->attributes['inn']        = trim($client_data->code); 
            $this->attributes['lastName']   = trim($client_data->lastName); 
            ...
        }
    }
```

You can use other ways to create custom attributes that you specified in `'model_data'` (`config/scoring.php`).

### Change log

Please see the [changelog](CHANGELOG.md) for more information on what has changed recently.

## Contributing

Please see [contributing.md](CONTRIBUTING.md) for details and a todolist.

### Security

If you discover any security related issues, please email author email instead of using the issue tracker.

## Credits

[Volodymyr Farylevych](https://github.com/arttiger)

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.



CHANGELOG.md
# Changelog

All notable changes to `laravel-scoring` will be documented in this file

### ver. 1.0.0 - 2020-05-13

- initial release;

### ver. 1.1.0 - 2020-05-17

- changed the method get report from UBKI;
- fixed bugs;


src/Scoring.php
<?php

namespace Arttiger\Scoring;

use Arttiger\Scoring\Factories\ProviderFactory;
use Arttiger\Scoring\Interfaces\Provider;
use Illuminate\Support\Facades\Config;

class Scoring
{
    protected $provider;
    protected $config;

    /**
     * Scoring constructor.
     *
     * @param Provider|null $provider
     */
    public function __construct(Provider $provider = null)
    {
        $this->config = (class_exists('Config') ? Config::get('scoring') : []);
        $this->provider = $provider;
    }

    /**
     * @param       $service
     * @param array $params
     *
     * @return mixed
     */
    public function scoring($service, $params = [])
    {
        $factory = new ProviderFactory();
        if ($this->provider = $factory->getProvider($service)) {
            return $this->provider->getScoring($params);
        }

        return false;
    }

    /**
     * @param       $service
     * @param array $params
     *
     * @return bool
     */
    public function status($service, $params = [])
    {
        $factory = new ProviderFactory();
        if ($this->provider = $factory->getProvider($service)) {
            return $this->provider->sendFeedback($params);
        }

        return false;
    }

    /**
     * @param       $service
     * @param array $params
     *
     * @return bool
     */
    public function ubkiReport($service, $params = [])
    {
        $factory = new ProviderFactory();
        if ($this->provider = $factory->getProvider($service)) {
            return $this->provider->getUbki($params);
        }

        return false;
    }

    /**
     * @param       $service
     * @param array $params
     *
     * @return mixed
     */
    public function pre_scoring($service, $params = [])
    {
        $factory = new ProviderFactory();
        if ($this->provider = $factory->getProvider($service)) {
            return $this->provider->getPreScoring($params);
        }

        return false;
    }
}


src/Providers/RiskTools.php
<?php

namespace Arttiger\Scoring\Providers;

use Arttiger\Scoring\Interfaces\Provider;
use GuzzleHttp\Client;
use GuzzleHttp\Psr7\Request;
use Illuminate\Support\Facades\Log;

class RiskTools implements Provider
{
    const URL_GET_SCORING = '/scoring/get_score'; // get scoring
    const URL_REPORT_UBKI = '/ubki/get_reports'; // get UBKI reports
    const URL_FEEDBACK = '/apps/upsert_bulk'; // updates of statuses of loans
    const URL_GET_PRE_SCORING = '/scoring/get_prescore'; // get pre-scoring

    protected $requestUrl;
    protected $secretKey;
    protected $params;
    protected $request;
    protected $mapScoring;
    protected $sync;

    protected $exeptions = [
        'ubki',
        'status_id',
        'closed_at',
        'amount_to_pay',
        'total_paid',
        'overdue_days',
    ];

    /**
     * RiskTools constructor.
     *
     * @param array $config
     */
    public function __construct(array $config)
    {
        if (! isset($config['url']) || $config['url'] === '') {
            throw new \InvalidArgumentException('URL must be string and not empty');
        }

        if (! isset($config['key']) || $config['key'] === '') {
            throw new \InvalidArgumentException('Key must be string and not empty');
        }
        $this->requestUrl = $config['url'] ?? null;
        $this->secretKey = $config['key'] ?? null;
        $this->sync = $config['sync'] ?? null;
        $this->mapScoring = $config['model_data'] ?? null;
    }

    /**
     * @param array $params
     *
     * @return array
     */
    public function getScoring(array $params = []): array
    {
        $this->params = $params;
        $this->request = self::requestScoring();

        return self::query(self::URL_GET_SCORING);
    }

    /**
     * @param array $params
     *
     * @return array
     */
    public function getPreScoring(array $params = []): array
    {
        $this->params = $params;
        $this->request = self::requestPreScoring();

        return self::query(self::URL_GET_PRE_SCORING);
    }

    /**
     * @param array $params
     *
     * @return array
     */
    public function sendFeedback(array $params = []): array
    {
        $this->params = $params;
        $this->request = self::requestFeedback();

        return self::query(self::URL_FEEDBACK);
    }

    /**
     * @param array $params
     *
     * @return array
     */
    public function getUbki(array $params = []): array
    {
        $this->params = $params;
        $this->request = self::requestGetUbki();

        return self::query(self::URL_REPORT_UBKI);
    }

    /**
     * @return string[]
     */
    private function headers()
    {
        return [
            'Content-type' => 'application/json',
            'AuthKey'      => $this->secretKey,
        ];
    }

    /**
     * Send a request to RiskTools and get a response.
     *
     * @param $method
     *
     * @return mixed
     */
    private function query($method)
    {
        if ($this->requestUrl && $this->request) {
            $client = new Client();
            $request = new Request(
                'POST',
                $this->requestUrl.$method,
                self::headers(),
                $this->request
            );

            try {
                $response = $client->send($request);
                if ($response->getStatusCode() == 200) {
                    $result = json_decode($response->getBody()->getContents(), true);
                    if ($result['status'] == 'error') {
                        Log::error('Response error', [static::class, $result]);
                    }

                    return $result;
                } else {
                    Log::error('No response to scoring request received.', [static::class, $response]);
                }
            } catch (\Exception $e) {
                $error = $e->getMessage();
                Log::error('Error GuzzleHttp Client.', [static::class, $error]);
            }
        }

        return [];
    }

    /**
     * @return false|string|void
     */
    private function requestScoring()
    {
        if (is_array($this->mapScoring)) {
            foreach ($this->mapScoring as $map_key => $map_value) {
                if (! in_array($map_key, $this->exeptions)) {
                    if (isset($this->params[$map_value])) {
                        $parameters[$map_key] = $this->params[$map_value];
                    } else {
                        Log::error('Mapping model does not contain an parameter (attribute): '.$map_value, [static::class]);
                    }
                }
            }
        } else {
            Log::error('Mapping model data is empty.', [static::class]);
        }

        $application['application'] = isset($parameters) ? $parameters : '';
        if (isset($this->params['ubki'])) {
            $application['ubki'] = $this->params['ubki'];
        }

        return json_encode($application);
    }

    /**
     * @return false|string|void
     */
    private function requestPreScoring()
    {
        $social_number = $this->params['social_number'];

        return json_encode(['social_number' => isset($social_number) ? $social_number : '']);
    }

    /**
     * @return false|string|void
     */
    private function requestFeedback()
    {
        if (is_array($this->mapScoring)) {
            foreach ($this->mapScoring as $map_key => $map_value) {
                if ($map_key != 'ubki') {
                    if (isset($this->params[$map_value])) {
                        $parameters[$map_key] = $this->params[$map_value];
                    }
                }
            }
        } else {
            Log::error('Mapping model data is empty.', [static::class]);
        }
        if ($this->sync) {
            $application['sync'] = $this->sync;
        }
        $application['applications'][] = isset($parameters) ? $parameters : '';

        return json_encode($application);
    }

    /**
     * @return false|string|void
     */
    private function requestGetUbki()
    {
        if (is_array($this->mapScoring)) {
            $map_value = $this->mapScoring['id'];
            if (isset($this->params[$map_value])) {
                $application_id = $this->params[$map_value];
            } else {
                Log::error('Mapping model does not contain an parameter (attribute): '.$map_value, [static::class]);
            }
        } else {
            Log::error('Mapping model data is empty.', [static::class]);
        }

        return json_encode(['application_id' => isset($application_id) ? $application_id : '']);
    }
}


src/Traits/ScoringTrait.php
<?php

namespace Arttiger\Scoring\Traits;

    use Arttiger\Scoring\Facades\Scoring;

    trait ScoringTrait
    {
        /**
         * Get scoring data.
         *
         * @param       $service
         * @param       $params
         *
         * @return mixed
         */
        public function scoring($service, $params = [])
        {
            if (method_exists($this, 'scoringAttributes')) {
                $this->scoringAttributes();
            }

            return Scoring::scoring($service, $this->getAttributes());
        }

        /**
         * Send status of credit.
         *
         * @param       $service
         * @param       $params
         *
         * @return mixed
         */
        public function status($service, $params = [])
        {
            if (method_exists($this, 'scoringAttributes')) {
                $this->scoringAttributes();
            }

            return Scoring::status($service, $this->getAttributes());
        }

        /**
         * Get UBKI reports.
         *
         * @param       $service
         * @param       $params
         *
         * @return mixed
         */
        public function ubkiReport($service, $params = [])
        {
            if (method_exists($this, 'scoringAttributes')) {
                $this->scoringAttributes();
            }

            return Scoring::ubkiReport($service, $this->getAttributes());
        }
    }


src/ScoringServiceProvider.php
<?php

namespace Arttiger\Scoring;

use Illuminate\Support\ServiceProvider;

class ScoringServiceProvider extends ServiceProvider
{
    /**
     * Bootstrap the application services.
     */
    public function boot()
    {
        $this->publishes([
            __DIR__.'/../config/scoring.php' => config_path('scoring.php'),
        ], 'config');

        // Publishing is only necessary when using the CLI.
        if ($this->app->runningInConsole()) {
            $this->bootForConsole();
        }
    }

    /**
     * Register the application services.
     */
    public function register()
    {
        $this->mergeConfigFrom(__DIR__.'/../config/scoring.php', 'scoring');

        // Register the service the package provides.
        $this->app->singleton('scoring', function ($app) {
            return new Scoring;
        });
    }

    /**
     * Get the services provided by the provider.
     *
     * @return array
     */
    public function provides()
    {
        return ['scoring'];
    }

    /**
     * Console-specific booting.
     *
     * @return void
     */
    protected function bootForConsole()
    {
        // Publishing the configuration file.
        $this->publishes([
            __DIR__.'/../config/scoring.php' => config_path('scoring.php'),
        ], 'config');
    }
}


config/scoring.php
<?php

    return [

        /*
        |--------------------------------------------------------------------------
        | Scoring service RiskTools
        |--------------------------------------------------------------------------
        |
        | Enable or disable scoring services are you using
        |
        */
        'risk_tools' => [
            'enabled'    => true,
            'url'        => env('RISK_TOOLS_URL', null),
            'key'        => env('RISK_TOOLS_KEY', null),
            'sync'       => env('RISK_TOOLS_SYNC', false),
            'class'      => Arttiger\Scoring\Providers\RiskTools::class,
            // Mapping model data (attributes) for get scoring
            'model_data' => [
                // For "scoring" method
                'id'              => 'public_id',     // ID of loan (string)
                'user_id'         => 'client_number', // ID of client (string)
                'social_number'   => 'inn',           // INN of client (string)
                'phone'           => 'phone',         // Phone of client (string)
                'email'           => 'email',         // Email of client
                'passport_number' => 'passport',      // Passport (example АА123456 or number ID-card)
                'passport_date'   => 'passport_date', // Passport issue date (example "2001-12-01")
                'first_name'      => 'firstName',     // First name
                'last_name'       => 'lastName',      // Last name
                'other_name'      => 'middleName',    // Middle name
                'birth_date'      => 'birth_date',    // Birth date (example "2001-12-01")
                'gender_id'       => 'gender_id',     // Gender (1 - male, 2 - female) (integer)
                'loan_amount'     => 'sum',           // Amount of credit (float)
                'loan_days'       => 'period',        // Number days of credit (integer)
                'applied_at'      => 'applied_at',    // Date and time created loan (example "2018-04-05T19:29:51+03:00")
                'ip'              => 'ip',            // IP of client
                'user_agent'      => 'user_agent',    // User-Agent of browser of client
                'ubki'            => 'ubki',          // XML-data from UBKI (array?, example ["<xml>", "<xml>"])

                // For "status" method
                // Status code of credit (1 - NEW, 2 - REJECT, 3 - APPROVED, 4 - ISSUED, 5 - CLOSED, 6 - OVERDUE)
                'status_id'       => 'status_code',   // (integer)
                'closed_at'       => 'closed_at',     // Date and time close loan (example "2018-04-25T10:05:00+03:00")
                'amount_to_pay'   => 'amount_debt',   // Total debt (float)
                'total_paid'      => 'total_paid',    // The sum of all payments (float)
                'overdue_days'    => 'overdue_days',  // The total number of days of delay of loan (integer)
            ],
        ],

        /*
        |--------------------------------------------------------------------------
        | Scoring service Scorista
        |--------------------------------------------------------------------------
        |
        | Enable or disable scoring services are you using
        |
        */
        'scorista'   => [
            'enabled' => false,
        ],

    ];
